### PR TITLE
[vulnerability] Add unordered transaction inputs

### PR DESCRIPTION
**Pull request by p-j01**
# Description 

From #2279, 

> The wallet uses ordered inputs of transactions (src/wallet/wallet.cpp, line 2628 - 2640), which may incur privacy risks like fingerprinting the wallet to observers.

This PR change the original order by shuffling the inputs.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2723,12 +2723,18 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
                 continue;
             }
         }
+        // Unordered transaction inputs
+        txNew.vin.clear();
+        std::vector<pair<const CWalletTx*,unsigned int>> selectedCoins(setCoins.begin(), setCoins.end());
+        random_shuffle(selectedCoins.begin(), selectedCoins.end(), GetRandInt);
```
Comment (fdoving) [inline src/wallet/wallet.cpp:2729]: random_shuffle() is obsoleted in c++17.
If that is not a problem, this looks good to me.

Comment (p-j01) [reply src/wallet/wallet.cpp:2729]: Okay, changing it for `std::shuffle`. We need a pseudo-random number engines and I am not sure if `std::default_random_engine` is enough ([C++ Reference](https://www.cplusplus.com/reference/random/)).

Comment (fdoving) [reply src/wallet/wallet.cpp:2729]: Could possibly backport Shuffle from bitcoin PR #14624

Comment (p-j01) [reply src/wallet/wallet.cpp:2729]: Might as well use `std::shuffle` with the `std::default_random_engine` with time-based seed in this PR and open a new issue for refactoring the shuffle function and the inputs format?

**Comment by patricklodder** [comment]
Merging this into 1.14.4 will immediately identify all 1.10.x and 1.14.x based wallets that may have vulnerabilities that we fix in 1.14.4 and any future releases. I would prefer to not merge this, but instead fix all the vulnerabilities as long as we have an active 1.14-maint, and only introduce this change in wallet behavior with 1.21. That way, we do not directly link wallet behavior to potential vulnerabilities.

Thoughts?

**Comment by p-j01** [comment]
Sure. Should we start merging them on `1.21-dev` (including some refactoring PRs)?

**Comment by patricklodder** [comment]
> Should we start merging them on 1.21-dev (including some refactoring PRs)?

Already part of 1.21-dev: https://github.com/dogecoin/dogecoin/commit/fd44ac1e8b

**Comment by p-j01** [comment]
+1

**Comment by rnicoll** [comment]
Restarting CI

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (rnicoll) [reply src/wallet/wallet.cpp]: I'm generally in favour of "If it improves the state of things, lets get it in, worry about perfect later" (but noting I have high standards for improving :) )

**Review by dgs2019 — APPROVED** [review]

**Review by dgs2019 — APPROVED** [review]

**Comment by rnicoll** [comment]
Ran a bundle of transactions with this on testnet, seems good as far as I can tell.

**Comment by patricklodder** [comment]
I'm still NAK on this because I heard no argument against my [assertion up there](https://github.com/dogecoin/dogecoin/pull/2314#issuecomment-864461832)

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (rnicoll) [reply src/wallet/wallet.cpp]: To clarify; I mean I like how this is, although targetting post 1.14.4

**Comment by ReverseControl** [comment]
@patricklodder I agree with you.



> Merging this into 1.14.4 will immediately identify all 1.10.x and 1.14.x based wallets that may have vulnerabilities that we fix in 1.14.4 and any future releases. I would prefer to not merge this, but instead fix all the vulnerabilities as long as we have an active 1.14-maint, and only introduce this change in wallet behavior with 1.21. That way, we do not directly link wallet behavior to potential vulnerabilities.
> 
> Thoughts?

**Pull request closed without merging**